### PR TITLE
Feature/java modules templates

### PR DESCRIPTION
**Pull request by guillermogutierrez**
**Why**
As part of the Java Stacks modularisation, each of the new six modules extracted from the existing templates will have their repo and build pipelines. All pipelines are identical except for some project-specific configurations. We decided to move all templates away from the java modules and put them in the pipeline-templates

**How**
Specific AzureDevOps templates for the java modules tasks to build and publish to artifactory are located under the folder:
- azDevOps
    |
    - azure
        |
        - templates
             |
             - steps
                |
                - java
Also new maven scripts were needed to support the publish and archetype requirements
  - scripts/build-maven-deploy.bash
  - scriptsbuild-maven-archetype.bash
  - scripts/build-maven-update-version.bash
All scripts have been implemented following the existing patterns and support parametrisation where possible.

### CODE REVIEW

**Review by russellseymour — APPROVED** [review]

File: scripts/build-maven-archetype.bash [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+	M2_LOCATION="./.m2"
+fi
+
+MAVEN_OPTIONS=" -Dmaven.repo.local=${M2_LOCATION} --no-transfer-progress -DpropertyFile=${ARCHETYPE_PROPERTIES_FILE}"
+
+if [ "${SETTINGS_LOCATION}" ]; then
+  MAVEN_OPTIONS+=" --settings ${SETTINGS_LOCATION} "
+fi
+
+./mvnw clean archetype:create-from-project ${MAVEN_OPTIONS}
```
Comment (ElvenSpellmaker) [inline scripts/build-maven-archetype.bash:64]: Should this script be in `stacks-pipeline-templates`?

File: scripts/build-maven-compile.bash [hunk 2]
```diff
@@ -47,7 +47,14 @@ if [ -z "${M2_LOCATION}" ]; then
...
+
+#./mvnw compile --no-transfer-progress -Dmaven.repo.local="${M2_LOCATION}" # TODO: Maybe this should run offline?
+#./mvnw process-test-resources --no-transfer-progress -Dmaven.repo.local="${M2_LOCATION}" --offline
+#./mvnw test-compile --no-transfer-progress -Dmaven.repo.local="${M2_LOCATION}"
+#./mvnw process-test-classes --no-transfer-progress -Dmaven.repo.local="${M2_LOCATION}" # TODO: Maybe this should be run offline?
+
+./mvnw compile ${MAVEN_OPTIONS} # TODO: Maybe this should run offline?
+./mvnw process-test-resources ${MAVEN_OPTIONS}
+./mvnw test-compile ${MAVEN_OPTIONS}
+./mvnw process-test-classes ${MAVEN_OPTIONS} # TODO: Maybe this should be run offline?
```
Comment (ElvenSpellmaker) [inline scripts/build-maven-compile.bash:60]: Should this script be in `stacks-pipeline-templates`?

File: scripts/build-maven-deploy.bash [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+if [ "${POM_FILE}" ]; then
+	MAVEN_OPTIONS+=" -f  ${POM_FILE} "
+fi
+
+if [ "${ALT_DEPLOYMENT_REPOSITORY}" ]; then
+	MAVEN_OPTIONS+=" -DaltDeploymentRepository=${ALT_DEPLOYMENT_REPOSITORY} "
+fi
+
+./mvnw deploy ${MAVEN_OPTIONS}
```
Comment (ElvenSpellmaker) [inline scripts/build-maven-deploy.bash:77]: Should this script be in `stacks-pipeline-templates`?

File: scripts/test-maven-generate-jacoco-report.bash [hunk 4]
```diff
@@ -46,7 +46,10 @@ if [ -z "$M2_LOCATION" ]; then
 	M2_LOCATION="./.m2"
 fi
 
+#./mvnw jacoco:report \
+#	--no-transfer-progress \
+#	-Dmaven.repo.local="${M2_LOCATION}" \
+#	--offline
```
Comment (ElvenSpellmaker) [inline scripts/test-maven-generate-jacoco-report.bash]: Is there a reason for the commented out code?

File: scripts/build-maven-archetype.bash [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+		exit 0
+	fi
+done
+
+while getopts "${OPTIONS}" option
+do
+	case "${option}" in
+		# Optional
+		A  ) ARCHETYPE_PROPERTIES_FILE="${OPTARG}";;
+    S  ) SETTINGS_LOCATION="${OPTARG}";;
```
Comment (ElvenSpellmaker) [inline scripts/build-maven-archetype.bash]: Indentation

File: scripts/build-maven-compile.bash [hunk 6]
```diff
@@ -47,7 +47,14 @@ if [ -z "${M2_LOCATION}" ]; then
...
-./mvnw compile --no-transfer-progress -Dmaven.repo.local="${M2_LOCATION}" # TODO: Maybe this should run offline?
-./mvnw process-test-resources --no-transfer-progress -Dmaven.repo.local="${M2_LOCATION}" --offline
-./mvnw test-compile --no-transfer-progress -Dmaven.repo.local="${M2_LOCATION}"
-./mvnw process-test-classes --no-transfer-progress -Dmaven.repo.local="${M2_LOCATION}" # TODO: Maybe this should be run offline?
+MAVEN_OPTIONS=" -Dmaven.repo.local=${M2_LOCATION} --no-transfer-progress "
+
+#./mvnw compile --no-transfer-progress -Dmaven.repo.local="${M2_LOCATION}" # TODO: Maybe this should run offline?
+#./mvnw process-test-resources --no-transfer-progress -Dmaven.repo.local="${M2_LOCATION}" --offline
+#./mvnw test-compile --no-transfer-progress -Dmaven.repo.local="${M2_LOCATION}"
+#./mvnw process-test-classes --no-transfer-progress -Dmaven.repo.local="${M2_LOCATION}" # TODO: Maybe this should be run offline?
```
Comment (ElvenSpellmaker) [inline scripts/build-maven-compile.bash]: Is there a reason for this commented out stuff?

File: scripts/build-maven-deploy.bash [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+do
+	case "${option}" in
+	  # Required
+	  u ) ARTIFACTORY_USER="${OPTARG}";;
+	  p ) ARTIFACTORY_PASSWORD="${OPTARG}";;
+
+		# Optional
+		R  ) ALT_DEPLOYMENT_REPOSITORY="${OPTARG}";;
+		F  ) POM_FILE="${OPTARG}";;
+    S  ) SETTINGS_LOCATION="${OPTARG}";;
```
Comment (ElvenSpellmaker) [inline scripts/build-maven-deploy.bash]: Indentation

File: scripts/build-maven-update-version.bash [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+		usage
+		exit 0
+	fi
+done
+
+while getopts "${OPTIONS}" option
+do
+	case "${option}" in
+		# Optional
+    V  ) PACKAGE_VERSION="${OPTARG}";;
```
Comment (ElvenSpellmaker) [inline scripts/build-maven-update-version.bash]: Indentation

**Comment by ElvenSpellmaker** [comment]
Generally LGTM, just a few minor things

File: scripts/build-maven-update-version.bash (hunk 8 above)
Comment (guillermogutierrez) [reply scripts/build-maven-update-version.bash]: Not sure why this is happening, file looks fine on my IDE

File: scripts/build-maven-deploy.bash (hunk 7 above)
Comment (guillermogutierrez) [reply scripts/build-maven-deploy.bash]: Not sure why this is happening, file looks fine on my IDE

File: scripts/build-maven-compile.bash (hunk 6 above)
Comment (guillermogutierrez) [reply scripts/build-maven-compile.bash]: Removed now,

File: scripts/build-maven-archetype.bash (hunk 5 above)
Comment (guillermogutierrez) [reply scripts/build-maven-archetype.bash]: Not sure why this is happening, file looks fine on my IDE

File: scripts/test-maven-generate-jacoco-report.bash (hunk 4 above)
Comment (guillermogutierrez) [reply scripts/test-maven-generate-jacoco-report.bash]: Removed now,

File: scripts/build-maven-update-version.bash (hunk 8 above)
Comment (ElvenSpellmaker) [reply scripts/build-maven-update-version.bash]: Tabs vs spaces, we've set tabs for Bash files, 

See if you can set your editor to show the difference between a tab and space maybe?
image

Comment (guillermogutierrez) [reply scripts/build-maven-update-version.bash]: Fixed

**Comment by ElvenSpellmaker** [comment]
LGTM :shipit:

**Pull request merged**